### PR TITLE
Fix 1423 Accessibility: The ToolStrip control UIA hierarchy of PrintPreviewDialog is not correct

### LINKED ISSUE #1423: Accessibility: The ToolStrip control UIA hierarchy of PrintPreviewDialog is not correct
Accessibility hierarchy of the PrintPreviewDialog's ToolStrip is not correct and may freeze the Inspect/other accessibility client app (due to infinite loop of UIA child elements):

image

### PR DESCRIPTION
**Pull request by M-Lipin**
Fixes #1423 

## Proposed changes

- Refactoring the ToolStripControlHost accessibility provider to make it reusable by derivates.
- Implement UIA Accessibility provider for ToolStrip-hosted NumericUpDown control and its host Too;Strip item.
- Rework other ToolStrip-hosted controls' accessibility where this is possible and safe.

### Before

Please see #1423 
- Incorrect position of spinner control in accessible hierarchy (should be the last element)
- Infinite loop of UIA elements (children of ToolStrip)

image

### After

image

## Test methodology 

- Manual testing
- Accessibility tools (Inspect, Narrator)
- UIA hierarchy shoul

## Accessibility testing  

- Without the change the accessible hierarchy of ToolStrip control in the PrintPreviewDialog is incorrect: Spinner control has incorrect position (first instead of last) and the collection of UIA elements is endless due to the loop between the first and the last element.

- With the change available accessible hierarchy is correct: spinner is the last child of ToolStrip, there is no infinite loop of ToolStrip children.

 

## Test environment(s) 

.NET Core SDK (reflecting any global.json):
Version: 3.0.100-preview8-012929
Commit: 795f8aeaa8
Runtime Environment:
OS Name: Windows
OS Version: 10.0.18362
OS Platform: Windows
RID: win10-x64
Base Path: C:\Program Files\dotnet\sdk\3.0.100-preview8-012929\

###### Microsoft Reviewers: [Open in CodeFlow](http://wpcp.azurewebsites.net/CodeFlowProtocolProxy2.php?pullrequest=https://github.com/dotnet/winforms/pull/1424)

### CODE REVIEW

File: src/System.Windows.Forms/src/System/Windows/Forms/Printing/PrintPreviewDialog.cs [hunk 1]
```diff
@@ -1283,10 +1284,10 @@ void InitForm()
             sixpagesToolStripButton = new System.Windows.Forms.ToolStripButton();
             separatorToolStripSeparator1 = new System.Windows.Forms.ToolStripSeparator();
             closeToolStripButton = new System.Windows.Forms.ToolStripButton();
-            pageCounter = new System.Windows.Forms.NumericUpDown();
+            pageCounterItem = new ToolStripNumericUpDown();
+            pageCounter = pageCounterItem.NumericUpDownControl;
             pageToolStripLabel = new System.Windows.Forms.ToolStripLabel();
             toolStrip1.SuspendLayout();
-            ((System.ComponentModel.ISupportInitialize)(pageCounter)).BeginInit();
```
Comment (RussKie) [inline src/System.Windows.Forms/src/System/Windows/Forms/Printing/PrintPreviewDialog.cs]: ❗️  If this is gone, `EndInit` should be gone too

File: src/System.Windows.Forms/src/System/Windows/Forms/ToolStrip.cs [hunk 2]
```diff
@@ -6192,5 +6192,70 @@ private RelativeLocation ComparePositions(Rectangle orig, Point check)
             return RelativeLocation.Left;
         }
     }
+
+    internal class ToolStripNumericUpDown : ToolStripControlHost
```
Comment (RussKie) [inline src/System.Windows.Forms/src/System/Windows/Forms/ToolStrip.cs]: Please move out to own file (e.g. `ToolStrip.ToolStripNumericUpDown`), this file is already huge as is

File: src/System.Windows.Forms/src/System/Windows/Forms/ToolStrip.cs [hunk 3]
```diff
@@ -6192,5 +6192,70 @@ private RelativeLocation ComparePositions(Rectangle orig, Point check)
...
+        public NumericUpDown NumericUpDownControl
+        {
+            get => _numericUpDownControl;
+        }
+
+        private static Control CreateControlInstance()
+        {
+            ToolStripNumericUpDownControl numericUpDown = new ToolStripNumericUpDownControl();
+            return numericUpDown;
+        }
```
Comment (RussKie) [inline src/System.Windows.Forms/src/System/Windows/Forms/ToolStrip.cs]: Why not
```
private static Control CreateControlInstance() => new ToolStripNumericUpDownControl();
```

File: src/System.Windows.Forms/src/System/Windows/Forms/ToolStripControlHost.cs [hunk 4]
```diff
@@ -1173,6 +1166,102 @@ internal override void OnKeyboardToolTipUnhook(ToolTip toolTip)
...
+            /// <summary>
+            /// Gets a description of the default action for an object.
+            /// </summary>
+            public override string DefaultAction
+            {
+                get
+                {
+                    // Note: empty value is provided due to this Accessible object
+                    // represents the control container but not the contained control
+                    // control itself.
```
Comment (RussKie) [inline src/System.Windows.Forms/src/System/Windows/Forms/ToolStripControlHost.cs]: Looks like a typo - `control` is used twice

**Comment by RussKie** [comment]
Please rebase on top of the latest master

File: src/System.Windows.Forms/src/System/Windows/Forms/Printing/PrintPreviewDialog.cs (hunk 1 above)
Comment (M-Lipin) [reply src/System.Windows.Forms/src/System/Windows/Forms/Printing/PrintPreviewDialog.cs]: Thanks!

Comment (M-Lipin) [reply src/System.Windows.Forms/src/System/Windows/Forms/Printing/PrintPreviewDialog.cs]: fixed

File: src/System.Windows.Forms/src/System/Windows/Forms/ToolStrip.cs (hunk 2 above)
Comment (M-Lipin) [reply src/System.Windows.Forms/src/System/Windows/Forms/ToolStrip.cs]: Done, thank you!

File: src/System.Windows.Forms/src/System/Windows/Forms/ToolStrip.cs (hunk 3 above)
Comment (M-Lipin) [reply src/System.Windows.Forms/src/System/Windows/Forms/ToolStrip.cs]: Fixed, thanks.

File: src/System.Windows.Forms/src/System/Windows/Forms/ToolStripControlHost.cs (hunk 4 above)
Comment (M-Lipin) [reply src/System.Windows.Forms/src/System/Windows/Forms/ToolStripControlHost.cs]: yes, thanks

Comment (M-Lipin) [reply src/System.Windows.Forms/src/System/Windows/Forms/ToolStripControlHost.cs]: fixed

File: src/System.Windows.Forms/src/System/Windows/Forms/ToolStripControlHost.cs [hunk 5]
```diff
@@ -1321,5 +1410,79 @@ void IDictionaryService.SetValue(object key, object value)
...
+                _toolStripHostedControl = toolStripHostedControl;
+            }
+
+            internal override UnsafeNativeMethods.IRawElementProviderFragmentRoot FragmentRoot
+            {
+                get
+                {
+                    if (_toolStripHostedControl != null // Hosted control should not be null.
+                        && _toolStripControlHost != null // ToolStripControlHost is a container for ToolStripControl.
+                        && _toolStripControlHost.Owner != null) { // Owner is the ToolStrip.
```
Comment (RussKie) [inline src/System.Windows.Forms/src/System/Windows/Forms/ToolStripControlHost.cs]: [nit] incorrect placement of the closing brace

File: src/System.Windows.Forms/src/System/Windows/Forms/ToolStripControlHost.cs [hunk 6]
```diff
@@ -1173,6 +1166,102 @@ internal override void OnKeyboardToolTipUnhook(ToolTip toolTip)
...
+
+            /// <summary>
+            /// Return the element that is the root node of this fragment of UI.
+            /// </summary>
+            internal override UnsafeNativeMethods.IRawElementProviderFragmentRoot FragmentRoot
+            {
+                get
+                {
+                    return _ownerItem.RootToolStrip.AccessibilityObject;
+                }
```
Comment (RussKie) [inline src/System.Windows.Forms/src/System/Windows/Forms/ToolStripControlHost.cs]: Could use an arrow function here

File: src/System.Windows.Forms/src/System/Windows/Forms/ToolStripControlHost.cs [hunk 7]
```diff
@@ -1173,6 +1166,102 @@ internal override void OnKeyboardToolTipUnhook(ToolTip toolTip)
...
+        protected override AccessibleObject CreateAccessibilityInstance()
+        {
+            return new ToolStripControlHostAccessibleObject(this);
+        }
+
+        /// <summary>
+        /// Defines the ToolStripControlHost AccessibleObject.
+        /// </summary>
+        [Runtime.InteropServices.ComVisible(true)]
+        internal class ToolStripControlHostAccessibleObject : ToolStripItemAccessibleObject
```
Comment (RussKie) [inline src/System.Windows.Forms/src/System/Windows/Forms/ToolStripControlHost.cs]: Does it have to be `internal`?

File: src/System.Windows.Forms/src/System/Windows/Forms/ToolStripControlHost.cs (hunk 5 above)
Comment (M-Lipin) [reply src/System.Windows.Forms/src/System/Windows/Forms/ToolStripControlHost.cs]: Fixed, thanks.

File: src/System.Windows.Forms/src/System/Windows/Forms/ToolStripControlHost.cs (hunk 6 above)
Comment (M-Lipin) [reply src/System.Windows.Forms/src/System/Windows/Forms/ToolStripControlHost.cs]: Yes, thanks.

File: src/System.Windows.Forms/src/System/Windows/Forms/ToolStripControlHost.cs (hunk 7 above)
Comment (M-Lipin) [reply src/System.Windows.Forms/src/System/Windows/Forms/ToolStripControlHost.cs]: Made it public. But can you please advise which approach is more preferable in whole: to make new definitions public in Core or to try preserve less new APIs?

Comment (M-Lipin) [reply src/System.Windows.Forms/src/System/Windows/Forms/ToolStripControlHost.cs]: Reverted to internal but need to investigate whether we can make it private.

**Pull request merged**